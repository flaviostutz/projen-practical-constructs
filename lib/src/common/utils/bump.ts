// import { Tasks } from "projen";

// export const bump = (project: Project, opts: BumpOptions) => {};
// parei aqui... fazer bump

// depois criar Tasks pra release