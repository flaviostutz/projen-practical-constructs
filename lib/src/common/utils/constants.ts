export const MONOTAG_VERSION = '1.23.0';
