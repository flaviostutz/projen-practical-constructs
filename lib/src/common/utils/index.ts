export * from './ts2jsii';
export * from './monotag';
