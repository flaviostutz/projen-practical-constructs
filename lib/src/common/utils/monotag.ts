import { NextTagOptions } from '../types/monotag';

export const monotagCliArgs = (opts: NextTagOptions): string => {
  const args: string[] = [];

  // add only args that were defined
  pushIfDefined(args, 'bump-action', opts.bumpAction);
  pushIfDefined(args, 'bump-files', opts.bumpFiles?.join(','));
  pushIfDefined(args, 'changelog-file', opts.changelogFile);
  pushIfDefined(args, 'from-ref', opts.fromRef);
  pushIfDefined(args, 'git-email', opts.gitEmail);
  pushIfDefined(args, 'git-username', opts.gitUsername);
  pushIfDefined(args, 'max-version', opts.maxVersion);
  pushIfDefined(args, 'min-version', opts.minVersion);
  pushIfDefined(args, 'notes-file', opts.notesFile);
  pushIfDefined(args, 'conv-commit', opts.onlyConvCommit ? 'true' : '');
  pushIfDefined(args, 'path', opts.path);
  pushIfDefined(args, 'prerelease', opts.preRelease ? 'true' : '');
  pushIfDefined(args, 'prerelease-increment', opts.preReleaseAlwaysIncrement ? 'true' : '');
  pushIfDefined(args, 'prerelease-id', opts.preReleaseIdentifier);
  pushIfDefined(args, 'repoDir', opts.repoDir);
  pushIfDefined(args, 'semver-level', opts.semverLevel);
  pushIfDefined(args, 'tag-file', opts.tagFile);
  pushIfDefined(args, 'tag-prefix', opts.tagPrefix);
  pushIfDefined(args, 'tag-suffix', opts.tagSuffix);
  pushIfDefined(args, 'to-ref', opts.toRef);
  pushIfDefined(args, 'verbose', opts.verbose ? 'true' : '');
  pushIfDefined(args, 'version-file', opts.versionFile);

  // add extra args
  if (opts.monotagExtraArgs) {
    // eslint-disable-next-line functional/immutable-data
    args.push(opts.monotagExtraArgs.trim());
  }
  return args.join(' ').trim();
};

const pushIfDefined = (args: string[], argName: string, argValue?: string): void => {
  if (argValue) {
    // eslint-disable-next-line functional/immutable-data
    args.push(argValue ? `--${argName}="${argValue}"` : '');
  }
};
