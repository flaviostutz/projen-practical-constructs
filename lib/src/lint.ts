/* eslint-disable no-restricted-syntax */
/* eslint-disable no-new */
import { Component, Project } from 'projen';

import { MyPy } from './components/mypy';
import { Ruff, RuffOptions } from './components/ruff';
import { PipAudit } from './components/pip-audit';

/**
 * Python project lint configurations
 */
export class LintTarget extends Component {
  constructor(project: Project, opts: LintOptions) {
    super(project);

    project.tasks.addTask('lint', {
      description: `Lint project (format, type check, audit, lint)`,
    });

    // MYPY - type checks
    new MyPy(project, { ...opts, attachTasksTo: 'lint' });

    // RUFF - linting and formatting
    new Ruff(project, { ...opts, attachTasksTo: 'lint' });

    // PIP-AUDIT - dependency vulnerability checks
    new PipAudit(project, { ...opts, attachTasksTo: 'lint' });
  }
}

export interface LintOptions extends RuffOptions {}
