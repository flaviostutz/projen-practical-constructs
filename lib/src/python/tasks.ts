import { Project, Task } from 'projen';

export interface TaskOptionsTarget {
  /**
   * Path to the python virtual environment directory
   * used in this project
   * @default .venv
   */
  readonly venvPath?: string;
  /**
   * Existing task to attach new tasks to.
   * It will be included as "spawn" tasks in new steps
   */
  readonly attachTasksTo?: string;
}

export interface TaskOptions {
  /**
   * Path to the python virtual environment directory
   * used in this project
   */
  readonly venvPath: string;
  /**
   * Existing task to attach new tasks to.
   * It will be included as "spawn" tasks in new steps
   */
  readonly attachTasksTo?: string;
}

export interface TaskOptionsWithFix extends TaskOptions {
  /**
   * Attach task to fix
   */
  readonly attachFixTasksTo?: string;
}

export const addTaskToParent = (project: Project, task: Task, parentTaskName?: string): void => {
  if (!parentTaskName) {
    return;
  }
  const attachTask = project.tasks.tryFind(parentTaskName);
  if (!attachTask) {
    throw new Error(`'${parentTaskName}' task not found`);
  }
  attachTask.spawn(task);
};
