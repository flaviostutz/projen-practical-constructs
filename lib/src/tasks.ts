import { Project, Task } from 'projen';

export interface TaskOptions {
  /**
   * Path to the python virtual environment directory
   * used in this project
   * @default .venv
   */
  readonly venvPath?: string;
  /**
   * Existing task to attach new tasks to.
   * It will be included as "spawn" tasks in new steps
   */
  readonly attachTasksTo?: string;
}

export const addTaskToParent = (project: Project, task: Task, parentTaskName?: string): void => {
  if (!parentTaskName) {
    return;
  }
  const attachTask = project.tasks.tryFind(parentTaskName);
  if (!attachTask) {
    throw new Error(`'${parentTaskName}' task not found`);
  }
  attachTask.spawn(task);
};
