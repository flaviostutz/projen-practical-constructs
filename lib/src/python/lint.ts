/* eslint-disable no-restricted-syntax */
/* eslint-disable no-new */
import { Component, Project } from 'projen';

import { MyPy } from './components/mypy';
import { Ruff, RuffOptions } from './components/ruff';
import { PipAudit } from './components/pip-audit';
import { TaskOptions } from './tasks';

/**
 * Python project lint configurations
 */
export class LintTarget extends Component {
  constructor(project: Project, taskOptions: TaskOptions, opts?: LintOptions) {
    super(project);

    // MYPY - type checks
    new MyPy(project, taskOptions);

    // RUFF - linting and formatting
    new Ruff(project, { ...taskOptions }, opts);

    // PIP-AUDIT - dependency vulnerability checks
    new PipAudit(project, taskOptions);
  }
}

export interface LintOptions extends RuffOptions {}
