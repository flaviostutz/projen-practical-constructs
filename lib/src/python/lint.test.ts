/* eslint-disable no-new */
import { Testing } from 'projen';

import { LintTarget } from './lint';
import { TaskOptions } from './tasks';
import { TestProject } from './utils';

test('LintTarget is synthesized correctly', () => {
  const project = new TestProject();
  const taskOpts: TaskOptions = { venvPath: '.venv' };
  new LintTarget(project, taskOpts);
  const output = Testing.synth(project);
  expect(output).toMatchSnapshot();
});
