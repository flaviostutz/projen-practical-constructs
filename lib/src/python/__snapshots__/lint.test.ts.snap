// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LintTarget is synthesized correctly 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

* text=auto eol=lf",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/mypy.ini
.mypy_cache
!/ruff.toml
.ruff_cache
.audit_cache
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "dependencies": [
      {
        "name": "mypy",
        "type": "devenv",
        "version": "1.14.*",
      },
      {
        "name": "pip-audit",
        "type": "devenv",
        "version": "2.*",
      },
      {
        "name": "ruff",
        "type": "devenv",
        "version": "0.8.*",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      "mypy.ini",
      "ruff.toml",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "lint": {
        "description": "Lint project (format, type check, audit, lint)",
        "name": "lint",
      },
      "lint-audit": {
        "description": "Code type checks (mypy)",
        "name": "lint-audit",
        "steps": [
          {
            "exec": ".venv/bin/pip-audit --cache-dir .cache/.audit_cache",
          },
        ],
      },
      "lint-fix": {
        "description": "Fix auto fixable lint issues",
        "name": "lint-fix",
        "steps": [
          {
            "spawn": "lint-ruff-fix",
          },
        ],
      },
      "lint-mypy": {
        "description": "Code type checks (mypy)",
        "name": "lint-mypy",
        "steps": [
          {
            "exec": ".venv/bin/mypy src tests",
          },
        ],
      },
      "lint-ruff": {
        "description": "Code checks (RUFF)",
        "name": "lint-ruff",
        "steps": [
          {
            "exec": ".venv/bin/ruff format --check src tests",
          },
          {
            "exec": ".venv/bin/ruff check src tests",
          },
        ],
      },
      "lint-ruff-fix": {
        "description": "Lint fix (RUFF)",
        "name": "lint-ruff-fix",
        "steps": [
          {
            "exec": ".venv/bin/ruff format src tests",
          },
          {
            "exec": ".venv/bin/ruff check --fix src tests",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "mypy.ini": "[mypy]
mypy_path = 
    src
    tests
warn_return_any = True
warn_unreachable = True
warn_redundant_casts = True
warn_unused_ignores = True
pretty = True
cache_dir = .cache/.mypy_cache
",
  "ruff.toml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

cache-dir = ".cache/.ruff_cache"
show-fixes = true
src = [ "src", "tests" ]
target-version = "py313"
unsafe-fixes = false

[format]
docstring-code-format = true

[lint]
ignore = [ ]
select = [ ]
per-file-ignores = { }

  [lint.mccabe]
  max-complexity = 14
",
}
`;
